File: platform-console/apps/raap-admin/src/api/core/prompt-optimizer.ts
import { requestClient } from '#/api/request';

export namespace PromptOptimizerApi {
  export type OptimizerMode =
    | 'batch_patch'
    | 'critic_patch'
    | 'global_refactor'
    | 'local_patch';

  export interface Patch {
    id: number;
    run_id: number;
    patch_index: number;
    operation: 'delete' | 'insert_after' | 'insert_before' | 'replace';
    old_text: string;
    new_text?: null | string;
    reason?: null | string;
    status: 'accepted' | 'edited' | 'pending' | 'rejected';
    edited_new_text?: null | string;
    review_comment?: null | string;
  }

  export interface Run {
    id: number;
    prompt_id: number;
    prompt_version_id: number;
    issue_id?: null | number;
    mode: OptimizerMode;
    model?: null | string;
    base_url?: null | string;
    temperature?: null | string;
    max_tokens?: null | number;
    status: 'failed' | 'pending' | 'running' | 'succeeded';
    input_snapshot?: null | Record<string, any>;
    raw_output?: null | string;
    parsed_output?: null | Record<string, any>;
    error_message?: null | string;
    patches: Patch[];
    create_time?: null | string;
    update_time?: null | string;
  }

  export interface CreateRunRequest {
    mode: OptimizerMode;
    problem_text: string;
    prompt_content?: string;
    prompt_name?: string;
    prompt_id?: number;
    prompt_version_id?: number;
    prompt_type?: 'critic' | 'generation' | 'other';
    generated_content?: string;
    generated_title?: string;
    model?: string;
    base_url?: string;
    temperature?: number;
    max_tokens?: number;
    timeout?: number;
    json_mode?: boolean;
    include_revised_prompt?: boolean;
  }

  export interface UpdatePatchRequest {
    status?: 'accepted' | 'edited' | 'pending' | 'rejected';
    edited_new_text?: string;
    review_comment?: string;
  }

  export interface ApplyPatchesRequest {
    patch_ids?: number[];
    change_summary?: string;
    created_by?: string;
    save_version?: boolean;
  }

  export interface ApplyPatchesResponse {
    applied_patch_ids: number[];
    conflicts: { patch_id: number; reason: string }[];
    candidate_content: string;
    new_version?: null | {
      id: number;
      prompt_id: number;
      version_no: number;
      content: string;
    };
  }
}

export async function createPromptOptimizerRunApi(
  data: PromptOptimizerApi.CreateRunRequest,
) {
  return requestClient.post<PromptOptimizerApi.Run>(
    '/v1/prompt-optimizer/runs',
    data,
  );
}

export async function updatePromptPatchApi(
  patchId: number,
  data: PromptOptimizerApi.UpdatePatchRequest,
) {
  return (requestClient as any).request(`/v1/prompt-optimizer/patches/${patchId}`, {
    data,
    method: 'PATCH',
  }) as Promise<PromptOptimizerApi.Patch>;
}

export async function applyPromptPatchesApi(
  runId: number,
  data: PromptOptimizerApi.ApplyPatchesRequest,
) {
  return requestClient.post<PromptOptimizerApi.ApplyPatchesResponse>(
    `/v1/prompt-optimizer/runs/${runId}/apply`,
    data,
  );
}

File: platform-server/app/schemas/prompt_optimizer.py
"""
Prompt optimizer schemas.
"""
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.base import BaseSchema, TimestampSchema


PromptType = Literal["generation", "critic", "other"]
IssueType = Literal["human_opinion", "review_problem", "batch_case"]
OptimizerMode = Literal["local_patch", "global_refactor", "critic_patch", "batch_patch"]
RunStatus = Literal["pending", "running", "succeeded", "failed"]
PatchOperation = Literal["replace", "delete", "insert_after", "insert_before"]
PatchStatus = Literal["pending", "accepted", "rejected", "edited"]


class PromptAssetCreate(BaseSchema):
    name: str = Field(..., max_length=255)
    content: str
    prompt_type: PromptType = "generation"
    tenant_code: Optional[str] = Field(default=None, max_length=64)
    description: Optional[str] = None
    tags: Optional[list[str]] = None
    created_by: Optional[str] = Field(default=None, max_length=64)


class PromptAssetUpdate(BaseSchema):
    name: Optional[str] = Field(default=None, max_length=255)
    prompt_type: Optional[PromptType] = None
    tenant_code: Optional[str] = Field(default=None, max_length=64)
    description: Optional[str] = None
    tags: Optional[list[str]] = None


class PromptVersionCreate(BaseSchema):
    content: str
    parent_version_id: Optional[int] = None
    source_run_id: Optional[int] = None
    change_summary: Optional[str] = None
    created_by: Optional[str] = Field(default=None, max_length=64)
    set_current: bool = True


class PromptAssetResponse(TimestampSchema):
    id: int
    tenant_code: Optional[str] = None
    name: str
    prompt_type: str
    description: Optional[str] = None
    current_version_id: Optional[int] = None
    tags: Optional[list[str]] = None
    is_deleted: int


class PromptVersionResponse(BaseSchema):
    id: int
    prompt_id: int
    version_no: int
    content: str
    parent_version_id: Optional[int] = None
    source_run_id: Optional[int] = None
    change_summary: Optional[str] = None
    created_by: Optional[str] = None
    create_time: Optional[datetime] = None


class PromptIssueResponse(BaseSchema):
    id: int
    prompt_id: int
    prompt_version_id: int
    issue_type: str
    problem_text: str
    generated_content: Optional[str] = None
    generated_title: Optional[str] = None
    issue_metadata: Optional[dict[str, Any]] = None
    create_time: Optional[datetime] = None


class PromptPatchResponse(TimestampSchema):
    id: int
    run_id: int
    patch_index: int
    operation: str
    old_text: str
    new_text: Optional[str] = None
    reason: Optional[str] = None
    status: str
    edited_new_text: Optional[str] = None
    review_comment: Optional[str] = None


class PromptOptimizerRunCreate(BaseSchema):
    mode: OptimizerMode
    problem_text: str
    prompt_id: Optional[int] = None
    prompt_version_id: Optional[int] = None
    prompt_content: Optional[str] = None
    prompt_name: Optional[str] = Field(default=None, max_length=255)
    prompt_type: PromptType = "generation"
    tenant_code: Optional[str] = Field(default=None, max_length=64)
    issue_type: Optional[IssueType] = None
    generated_content: Optional[str] = None
    generated_title: Optional[str] = None
    issue_metadata: Optional[dict[str, Any]] = None
    model: Optional[str] = None
    base_url: Optional[str] = None
    api_key: Optional[str] = None
    temperature: float = 0.2
    max_tokens: int = 8000
    timeout: int = 90
    json_mode: bool = True
    include_revised_prompt: bool = False


class PromptOptimizerRunResponse(TimestampSchema):
    id: int
    prompt_id: int
    prompt_version_id: int
    issue_id: Optional[int] = None
    mode: str
    model: Optional[str] = None
    base_url: Optional[str] = None
    temperature: Optional[str] = None
    max_tokens: Optional[int] = None
    status: str
    input_snapshot: Optional[dict[str, Any]] = None
    raw_output: Optional[str] = None
    parsed_output: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    patches: list[PromptPatchResponse] = []


class PromptPatchUpdate(BaseSchema):
    status: Optional[PatchStatus] = None
    edited_new_text: Optional[str] = None
    review_comment: Optional[str] = None


class PromptPatchApplyRequest(BaseSchema):
    patch_ids: Optional[list[int]] = None
    change_summary: Optional[str] = None
    created_by: Optional[str] = Field(default=None, max_length=64)
    save_version: bool = True


class PromptPatchApplyConflict(BaseModel):
    patch_id: int
    reason: str


class PromptPatchApplyResponse(BaseSchema):
    applied_patch_ids: list[int]
    conflicts: list[PromptPatchApplyConflict]
    candidate_content: str
    new_version: Optional[PromptVersionResponse] = None
